File: sqlite2mysql/main.js
const mysql = require('mysql2');
const Database = require('better-sqlite3');
const db = new Database('./ruten.sqlite');
const fs = require('fs').promises;
let category_id = 100018;
let fenzu = '淘幂铺';
(async () => {
    // MySQL connection
    const mysqlDb = mysql.createConnection({
        host: '192.168.1.30',
        user: 'lutian',
        password: 'xdWETicFHZFf3bcT',
        database: 'lutian',
        charset: 'utf8mb4',
    }).promise();


    try {
        const stmt = db.prepare(`SELECT * FROM itemview where 分組='${fenzu}'`);
        await mysqlDb.beginTransaction();
        let jsonString=""
        // 使用 iterate 方法逐行讀取
        for (const row of stmt.iterate()) {
            console.log('Row:', row);
            let images = row.圖片.replace(/\\/g, '/').replace(/C:\/lutian\/PicBackup/g, 'https://img.1pan.me/photos').replace(/\/PicBackup/g, 'https://img.1pan.me/photos');
            images=images.replace(/D:\/露天自動上架精靈/g,'')
            images = images.split('|');
            jsonString = JSON.stringify(images);
            let tempstr = "";
            let skutemp = [];
            let escapedData="";
            try {
                try {
                    if (row.自訂規格 != "") {
                        row.自訂規格=row.自訂規格.replace(/\'/g, '"');
                        const skuType = JSON.parse(row.自訂規格)[0];
                        skuType.dataRows.forEach((item) => {
                            const skulist = [];
                            let temp = {
                                "name": {
                                    "zh_hk": item.name
                                },
                                "values": [],
                                "isImage": false
                            }
                            for (let index = 0; index < item.specValue.length; index++) {
                                const element = item.specValue[index];
                                let obj = {
                                    name: {
                                        zh_hk: element.name,
                                    }
                                };
                                if (element.simage != null) {
                                    obj.image = element.simage;
                                    if (obj.image.startsWith('http') == false) {
                                        obj.image = `https:${obj.image}`;
                                    }
                                    temp.isImage = true;
                                }
                                skulist.push(obj);
                            }
                            temp.values = skulist;
                            skutemp.push(temp);
                        })

                        tempstr = JSON.stringify(skutemp);
                    }else{
                        tempstr = "[]";
                    }
                } catch (e) {
                    throw e;
                }
                try {
                    let deschtml = row.說明.replace(/\\/g, '/')
                    let html = await fs.readFile(`/Users/forke/Downloads/trade${deschtml}`, 'utf8')
                    // escapedData = mysql.escape(html);
                    escapedData = html;
                } catch (e) {
                    throw e;
                }
                let price = row.直購價 || row.起標價;
                price=parseFloat(price)
                let [rows, fields] = await mysqlDb.query(`INSERT INTO products (price, images, variables,brand_id,video) VALUES (?, ?, ?,0,"")`, [price, jsonString, tempstr]);

                console.log('Inserted:', rows.insertId);
                productid = rows.insertId;

                if (skutemp.length != 0) {
                    const skuPrice = JSON.parse(row.自訂規格)[1];
                    let result = combineNames(skutemp.map(item => item.values));
                    let result1 = combineIndexes(skutemp.map(item => item.values));
                    for (let index = 0; index < result.length; index++) {
                        const element = result[index];
                        let priceObj = skuPrice[element];
                        if (priceObj) {
                            let price = priceObj.price
                            let is_default = index == 0 ? 1 : 0;
                            await mysqlDb.query(`INSERT INTO product_skus (product_id, variants, position,sku,price,origin_price,cost_price,quantity,is_default) VALUES (?, ?, ?,?,?,?,?,?,?)`, [productid, JSON.stringify(result1[index]), index, rand(), price, price, price, 1000, is_default]);
                        }
                    }
                } else {
                    await mysqlDb.query(`INSERT INTO product_skus (product_id, variants, position,sku,price,origin_price,cost_price,quantity,is_default) VALUES (?, ?, ?,?,?,?,?,?,?)`, [productid, `""`, 0, rand(), price, price, price, 1000, 1]);
                }

                await mysqlDb.query(`INSERT INTO product_categories (product_id, category_id) VALUES (?, ?)`, [productid, category_id]);

                await mysqlDb.query(`INSERT INTO product_descriptions (product_id, locale, name,content) VALUES (?, ?, ?,?)`, [productid, 'zh_hk', row.標題, escapedData]);
            } catch (e) {
                fs.appendFile('error.txt', `${row.ID}\n`);
                console.log('Row:', e);
                throw e;
            }
        }
        await mysqlDb.commit();

        console.log('事务已提交');
    } catch (e) {
        await mysqlDb.rollback();
        console.error('事务回滚:', e);
        throw e
    }
    process.on('exit', () => {
        sqliteDb.close();
        mysqlDb.end();
    });

})();

function combineNames(arrays) {
    const result = [];

    function backtrack(path, depth) {
        if (depth === arrays.length) {
            result.push(path.join("/"));
            return;
        }

        for (let i = 0; i < arrays[depth].length; i++) {
            let str = arrays[depth][i].name.zh_hk;
            path.push(`${str}`);  // 将索引转为字符串
            backtrack(path, depth + 1);
            path.pop();
        }
    }

    backtrack([], 0);
    return result;
}
function combineIndexes(arrays) {
    const result = [];

    function backtrack(path, depth) {
        if (depth === arrays.length) {
            result.push([...path]);
            return;
        }

        for (let i = 0; i < arrays[depth].length; i++) {
            path.push(`${i}`);  // 将索引转为字符串
            backtrack(path, depth + 1);
            path.pop();
        }
    }

    backtrack([], 0);
    return result;
}
function rand() {
    const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    const numbers = "0123456789";

    let result = "";

    // 生成5个随机字母
    for (let i = 0; i < 5; i++) {
        result += letters.charAt(Math.floor(Math.random() * letters.length));
    }

    // 生成10个随机数字
    for (let i = 0; i < 10; i++) {
        result += numbers.charAt(Math.floor(Math.random() * numbers.length));
    }

    return result;
}

File: sqlite2mysql/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"

	"github.com/tidwall/gjson"
	"github.com/tidwall/sjson"
	"gorm.io/driver/mysql"
	"gorm.io/driver/sqlite"
	"gorm.io/gorm"
)

type SKU struct {
	Image string            `json:"image"`
	Name  map[string]string `json:"name"`
}

var temp = `[
    {
        "name": {
            "zh_hk": "規格",
            "zh_cn": "規格"
        },
        "values": [],
		"isImage": false
}`

// 定义结构体
type SQLiteProduct struct {
	ID          uint    `gorm:"primaryKey"`
	Title       string  `gorm:"column:標題"`
	Description string  `gorm:"column:說明"`
	Price       float64 `gorm:"column:直購價"`
	Image       string  `gorm:"column:圖片"`
	CustomSpec  string  `gorm:"column:自訂規格"`
}

func (SQLiteProduct) TableName() string {
	return "itemview"
}

type MySQLProduct struct {
	ID         uint    `gorm:"primaryKey"`
	Price      float64 `gorm:"column:price"`
	Image      string  `gorm:"column:images"`
	CustomSpec string  `gorm:"column:variables"`
}

func (MySQLProduct) TableName() string {
	return "products_1"
}
func main() {
	// 连接 SQLite 数据库
	sqliteDB, err := gorm.Open(sqlite.Open("../ruten.sqlite"), &gorm.Config{})
	if err != nil {
		log.Fatalf("Failed to connect to SQLite: %v", err)
	}

	// 连接 MySQL 数据库
	dsn := "sql_192_168_1_30:3839b7857eb1f@tcp(192.168.1.30:3306)/sql_192_168_1_30?charset=utf8mb4&parseTime=True&loc=Local"
	mysqlDB, err := gorm.Open(mysql.Open(dsn), &gorm.Config{})
	if err != nil {
		log.Fatalf("Failed to connect to MySQL: %v", err)
	}

	// 确保 MySQL 表存在
	err = mysqlDB.AutoMigrate(&MySQLProduct{})
	if err != nil {
		log.Fatalf("Failed to migrate MySQL schema: %v", err)
	}
	var count int64
	if err := sqliteDB.Table("itemview").Count(&count).Error; err != nil {
		log.Fatalf("Failed to count records from itemview: %v", err)
	}

	log.Printf("Total records in a.sqlite: %d", count)
	// 逐行處理數據
	rows, err := sqliteDB.Table("itemview").Rows()
	if err != nil {
		log.Fatalf("Failed to fetch data from itemview: %v", err)
	}
	defer rows.Close()

	for rows.Next() {
		var item SQLiteProduct
		if err := sqliteDB.ScanRows(rows, &item); err != nil {
			log.Printf("Failed to scan row: %v", err)
			continue
		}

		// 处理图片字段
		item.Image = strings.ReplaceAll(item.Image, "\\", "/")
		item.Image = strings.ReplaceAll(item.Image, `C:/lutian/PicBackup`, "")
		images := strings.Split(item.Image, "|")
		jsonData, err := json.Marshal(images)
		if err != nil {
			fmt.Println("Error serializing slice:", err)
			return
		}

		// 转换为字符串输出
		jsonString := string(jsonData)

		skus := gjson.Get(item.CustomSpec, "1").Map()
		skulist := []SKU{}
		for k, _ := range skus {
			lang := map[string]string{"zh_hk": "規格", "zh_cn": "规格"}
			lang["zh_cn"] = k
			lang["zh_hk"] = k
			skulist = append(skulist, SKU{Name: lang, Image: ""})
		}
		value, _ := sjson.Set(temp, "values", skulist)
		fmt.Println(value)
		product := MySQLProduct{
			Price:      item.Price,
			Image:      jsonString,
			CustomSpec: value,
		}

		if err := mysqlDB.Create(&product).Error; err != nil {
			log.Printf("Failed to insert product into MySQL: %v", err)
		}
		log.Printf("Inserted product: %s", product.ID)
	}

	fmt.Println("Data migration completed successfully!")
}
